src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java

```diff
@@ -579,28 +579,65 @@ private static String getStringInsideTag(String result, DetailNode detailNode) {
      * Finds and returns first sentence.
      *
      * @param ast Javadoc root node.
+     * @param period Period character.
      * @return first sentence.
      */
-    private static String getFirstSentence(DetailNode ast) {
+    private static String getFirstSentence(DetailNode ast, String period) {
+        final Deque<DetailNode> stack = new LinkedList<>();
+        stack.push(ast);
         final StringBuilder result = new StringBuilder(256);
-        final String periodSuffix = DEFAULT_PERIOD + ' ';
-        for (DetailNode child : ast.getChildren()) {
-            final String text;
-            if (child.getChildren().length == 0) {
-                text = child.getText();
+        boolean foundPeriod = false;
+        while (!stack.isEmpty()) {
+            final DetailNode node = stack.pop();
+            if (node.getChildren().length == 0
+                && appendUpToSentenceEndingPeriod(node.getText(), period, result)) {
+                foundPeriod = true;
+                break;
             }
-            else {
-                text = getFirstSentence(child);
+            // Pushing last child first means it will be processed last
+            for (int childIndex = node.getChildren().length - 1; childIndex >= 0; childIndex--) {
+                stack.push(node.getChildren()[childIndex]);
             }
+        }
+        if (!foundPeriod) {
+            result.setLength(0);
+        }
+        return result.toString();
+    }
 
-            if (text.contains(periodSuffix)) {
-                result.append(text, 0, text.indexOf(periodSuffix) + 1);
+    /**
+     * Find the end of a sentence, and append the sentence to the result if it ends with a period.
+     * If no period is present, append the whole string to the result. The end of sentence detection
+     * here could be replaced in the future by Java's built-in BreakIterator class.
+     *
+     * @param text string to append to result
+     * @param period period character to find
+     * @param result builder to append to
+     * @return true if a sentence ending period was found, false otherwise
+     */
+    private static boolean appendUpToSentenceEndingPeriod(
+            String text, String period, StringBuilder result) {
+        int periodIndex = text.indexOf(period);
+        boolean foundPeriod = false;
+        while (periodIndex >= 0) {
+            final int afterPeriodIndex = periodIndex + period.length();
+
+            // Handle western period separately as it is only the end of a sentence if followed
+            // by whitespace. Other period characters often include whitespace in the character.
```

Comment:
Please help us to improve public doc on property period https://checkstyle.org/checks/javadoc/summaryjavadoc.html#Properties
To mention this nuance.

Reply:
This is last review item.

Reply:
I've been trying to figure out how the documentation works. It looks like it's generated with Doxia, but there's two XML files checked in:

https://github.com/checkstyle/checkstyle/blob/master/src/xdocs/checks/javadoc/summaryjavadoc.xml
https://github.com/checkstyle/checkstyle/blob/master/src/xdocs/checks/javadoc/summaryjavadoc.xml.template

I can see there's a macro in the template that seems to pull the Javadoc out of the check class. Is the idea maybe to update the Javadoc, run the templating and then manually edit its output? How do I use it?

Reply:
Sorry that I didn't guide you.
We just need update 
https://github.com/checkstyle/checkstyle/blob/ad6bfa80aced6ef6ef7e747fd6cc75c68256dba1/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/SummaryJavadocCheck.java#L51
And duplicate work at setter of this method. We did not finish automation for single place.
Please do not hesitate to pin me for any questions.

Reply:
Thanks, I've changed it. How does that look?

Reply:
Wording is good, let's make CI green to let me approve PR